Cassandra: rewrite on top of Java driver 4.6.0 (#2182)
akka/alpakka: docs/src/main/paradox/cassandra.md

  group=com.lightbend.akka
  artifact=akka-stream-alpakka-cassandra_$scala.binary.version$
  version=$project.version$
}

The table below shows direct dependencies of this module and the second tab shows all libraries it depends on transitively.

@@dependencies { projectId="cassandra" }

## Sessions

Cassandra is accessed through @apidoc[CassandraSession]s which are managed by the @apidoc[CassandraSessionRegistry] Akka extension. This way a session is shared across all usages within the actor system and properly shut down after the actor system is shut down.

> @scala[The `CassandraSession` is provided to the stream factory methods as `implicit` parameter.]

Reviewer: ```suggestion
@scala[The `CassandraSession` is provided to the stream factory methods as an `implicit` parameter.]
```

Scala
: @@snip [snip](/cassandra/src/test/scala/docs/scaladsl/CassandraSourceSpec.scala) { #init-session }

Java
: @@snip [snip](/cassandra/src/test/java/docs/javadsl/CassandraSourceTest.java) { #init-session }

See @ref[custom session creation](#custom-session-creation) below for tweaking this.

## Reading from Cassandra

> @apidoc[CassandraSource] provides factory methods to get Akka Streams Sources from CQL queries and from statements.

Reviewer: ```suggestion
@apidoc[CassandraSource] provides factory methods to get Akka Streams Sources from CQL queries and from @javadoc[com.datastax.oss.driver.api.core.cql.Statement](com.datastax.oss.driver.api.core.cql.Statement)s.
```

Dynamic parameters can be provided to the CQL as variable arguments.

Scala
: @@snip [snip](/cassandra/src/test/scala/docs/scaladsl/CassandraSourceSpec.scala) { #cql }

Java
: @@snip [snip](/cassandra/src/test/java/docs/javadsl/CassandraSourceTest.java) { #cql }

If the statement requires specific settings, you may pass any @javadoc[com.datastax.oss.driver.api.core.cql.Statement](com.datastax.oss.driver.api.core.cql.Statement).

Scala
: @@snip [snip](/cassandra/src/test/scala/docs/scaladsl/CassandraSourceSpec.scala) { #statement }

Java
: @@snip [snip](/cassandra/src/test/java/docs/javadsl/CassandraSourceTest.java) { #statement }

> Here we used a basic sink to complete the stream by collecting all of the stream elements to a collection. The power of streams comes from building larger data pipelines which leverage backpressure to ensure efficient flow control. Feel free to edit the example code and build @extref:[more advanced stream topologies](akka:stream/stream-introduction.html).

Reviewer: ```suggestion
Here we used a basic sink to complete the stream by collecting all of the stream elements into a collection. The power of streams comes from building larger data pipelines which leverage backpressure to ensure efficient flow control. Feel free to edit the example code and build @extref:[more advanced stream topologies](akka:stream/stream-introduction.html).
```

## Updating Cassandra

> @apidoc[CassandraFlow] provides factory methods to get Akka Streams flows to run CQL update statements. Alpakka Cassandra creates a @javadoc[PreparedStatement](com.datastax.oss.driver.api.core.cql.PreparedStatement) and for every stream element the `statementBinder` function binds the CQL placeholders to data.

Reviewer: This makes it sound that only literal "update" statements are supported.  I assume you can insert or upsert as well.  Maybe this section could be rephrased to "Writing to Cassandra".
